### PR TITLE
refresh favorites

### PR DESCRIPTION
**Pull request by kimyen**
In Progress

### CODE REVIEW

File: src/main/java/org/sagebionetworks/web/client/widget/header/Header.java [hunk 1]
```diff
@@ -50,32 +57,45 @@ public Widget asWidget() {
...
+				@Override
+				public void invoke() {
+					// only get called on failure
+					view.clearFavorite();
+				}
+			});
+		}
 		view.refresh();
 		view.setSearchVisible(true);
 	}
```
Comment (kimyen) [inline src/main/java/org/sagebionetworks/web/client/widget/header/Header.java:82]: Please review refresh()
case 1, both cachedUserSessionData & userSessionData are null => view.setUser()
case 2, cachedUserSessionData is null and userSessionData is not => view.setUser() then initUserFavorites()
case 3, cachedUserSessionData is not null and userSessionData is null => view.setUser() then do not initUserFavorites()
case 4, both are not null & different => view.setUser() then initUserFavorites()
case 5, both are not null & same => initUserFavorites()

File: src/main/java/org/sagebionetworks/web/client/widget/header/Header.java [hunk 2]
```diff
@@ -103,4 +125,19 @@ public void onFavoriteClick() {
...
+			@Override
+			public void onSuccess(List<EntityHeader> favorites) {
+				globalApplicationState.setFavorites(favorites);
+			}
+			@Override
+			public void onFailure(Throwable caught) {
+				callback.invoke();
+			}
+		});
+	}
```
Comment (kimyen) [inline src/main/java/org/sagebionetworks/web/client/widget/header/Header.java:141]: should initUserFavorites() be private?

File: src/main/java/org/sagebionetworks/web/client/widget/header/HeaderViewImpl.java [hunk 3]
```diff
@@ -241,21 +240,12 @@ public void setMenuItemActive(MenuItems menuItem) {
 	public void removeMenuItemActive(MenuItems menuItem) {
 	}
 
-	private void refreshTestSiteHeader() {
-		testSitePanel.setVisible(DisplayUtils.isInTestWebsite(cookies));
-	}
-
```
Comment (kimyen) [inline src/main/java/org/sagebionetworks/web/client/widget/header/HeaderViewImpl.java:247]: moved this down together with other private methods

File: src/main/java/org/sagebionetworks/web/client/widget/header/Header.java [hunk 4]
```diff
@@ -50,32 +57,45 @@ public Widget asWidget() {
...
-	
+
 	public void setSearchVisible(boolean searchVisible) {
 		view.setSearchVisible(searchVisible);
 	}
-	
+
 	public void refresh() {
+		UserSessionData userSessionData = authenticationController.getCurrentUserSessionData();
+		if (cachedUserSessionData == null || cachedUserSessionData.equals(userSessionData)) {
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/widget/header/Header.java]: Should only get into if null or user change.  Should contain init favorites call (not get favorites on every refresh).

File: src/main/java/org/sagebionetworks/web/client/widget/header/Header.java (hunk 2 above)
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/widget/header/Header.java:141]: Leave public for test cases

File: src/test/java/org/sagebionetworks/web/unitclient/widget/header/HeaderTest.java [hunk 5]
```diff
@@ -134,4 +139,34 @@ public void testOnFavoriteClickNonEmptyCase() {
...
+		// After User Logged in
+		UserSessionData userSessionData = new UserSessionData();
+		when(mockAuthenticationController.getCurrentUserSessionData()).thenReturn(userSessionData);
+		header.refresh();
+		verify(mockSynapseClient).getFavorites(any(AsyncCallback.class));
+
+		// User refresh
+		when(mockAuthenticationController.getCurrentUserSessionData()).thenReturn(userSessionData);
+		header.refresh();
+		verify(mockSynapseClient, Mockito.times(2)).getFavorites(any(AsyncCallback.class));
```
Comment (jay-hodgson) [inline src/test/java/org/sagebionetworks/web/unitclient/widget/header/HeaderTest.java]: 1 time

**Pull request merged**